src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,731 @@
+        OperatorMagnitudeInfo memory info = operatorMagnitudeInfo[operator][strategy];
+        
+        (uint64 freeMagnitudeToAdd, uint192 completed) =
+            _getPendingFreeMagnitude(operator, strategy, numToComplete, info.nextPendingFreeMagnitudeIndex);
+        info.freeMagnitude += freeMagnitudeToAdd;
+        info.nextPendingFreeMagnitudeIndex += completed;
+
+        operatorMagnitudeInfo[operator][strategy] = info;
+    }
+
+    /**
+     * @notice For a single strategy, modify magnitude allocations for each of the specified operatorSets
+     * @param operator address to modify allocations for
+     * @param allocation the magnitude allocations to modify for a single strategy
+     * @param allocationEffectTimestamp the timestamp when the allocations will take effect
+     * @param deallocationCompletableTimestamp the timestamp when the deallocations will be completable
+     * @dev For each allocation, allocation.operatorSets MUST be ordered in ascending order according to the
+     * encoding of the operatorSet. This is to prevent duplicate operatorSets being passed in. The easiest way to ensure
+     * ordering is to sort allocated operatorSets by address first, and then sort for each avs by ascending operatorSetIds.
+     */
+    function _modifyAllocations(
+        address operator,
+        MagnitudeAllocation calldata allocation,
+        uint32 allocationEffectTimestamp,
+        uint32 deallocationCompletableTimestamp
+    ) internal {
+        require(allocation.operatorSets.length == allocation.magnitudes.length, InputArrayLengthMismatch());
+
+        for (uint256 i = 0; i < allocation.operatorSets.length; ++i) {
+            require(
+                avsDirectory.isOperatorSet(allocation.operatorSets[i].avs, allocation.operatorSets[i].operatorSetId),
+                InvalidOperatorSet()
+            );
+
+            bytes32 operatorSetKey = _encodeOperatorSet(allocation.operatorSets[i]);
+            // prep magnitudeUpdates in storage
+
+            // Check that there is at MOST `MAX_PENDING_UPDATES` combined allocations & deallocations for the operator, operatorSet, strategy
+            (uint256 numPendingAllocations, uint256 numPendingDeallocations) =
+                _getPendingAllocationsAndDeallocations(operator, allocation.strategy, operatorSetKey);
+
+            require(
+                numPendingAllocations + numPendingDeallocations < MAX_PENDING_UPDATES,
+                PendingAllocationOrDeallocation()
+            );
+
+            Snapshots.History storage magnitudeUpdates = _magnitudeUpdate[operator][allocation.strategy][operatorSetKey];
+
+            // Read current magnitude allocation including its respective array index and length.
+            // We'll use these values later to check the number of pending allocations/deallocations.
+            (uint224 currentMagnitude,,) =
+                magnitudeUpdates.upperLookupRecentWithPos(uint32(block.timestamp));
```

Comment:
We literally do this twice in a row. The `_getPending` helper above calls this exact function the same way.

Reply:
And it's not even necessary.

If we separate out pending alloc/dealloc into a separate singular state variable, then to fetch the `currentMagnitude` here all we need to do is read the final element in `magnitudeUpdates`. We don't need to do `upperLookupRecentWithPos`.

Reply:
Fixed with Yash's refactor